Passivity indices and support for discrete time systems.

Mark-Yeatman opened the pull request:
This adds support for input and output passivity indices. The intention is to mimic the functionality of [ispassive](https://www.mathworks.com/help/control/ref/lti.ispassive.html) and [getPassiveIndex](https://www.mathworks.com/help/control/ref/lti.getpassiveindex.html) from Matlab (this implementation ignores frequency bands, relative passivity index, combined I/O passivity, and directional passivity index.)

This also adds support for computing indices and "is passive or not" for discrete time systems.

Mark-Yeatman commented:
Seems like the pipeline is having some issues unrelated to my changes?

bnavigator commented:
Yes it's unrelated. I'll have a look shortly.

bnavigator commented on control/passivity.py:
```diff
@@ -1,41 +1,68 @@
         raise ModuleNotFoundError("cvxopt required for passivity module")
 
+    if sys.ninputs != sys.noutputs:
+        raise Exception(
```
Please use a more non-generic `Exception` class like `ControlDimension`

Mark-Yeatman replied:
Change taken.

bnavigator commented on control/passivity.py:
```diff
@@ -1,41 +1,68 @@
+            "The number of system inputs must be the same as the number of system outputs.")
+
+    if rho is None and nu is None:
+        raise Exception("rho or nu must be given a float value.")
```
`control.exceptions.ControlArgument`

Mark-Yeatman replied:
Change taken.

bnavigator commented on control/tests/passivity_test.py:
```diff
@@ -2,16 +2,17 @@
 Author: Mark Yeatman  
 Date: May 30, 2022
 '''
+from tokenize import Double
```
?

Mark-Yeatman replied:
I also blame vscode here 🤷‍♂️

bnavigator commented on control/tests/rlocus_test.py:
```diff
@@ -112,7 +112,7 @@ def test_root_locus_zoom(self):
         assert_array_almost_equal(zoom_y, zoom_y_valid)
 
-    @pytest.mark.timeout(2)
+    # @pytest.mark.timeout(2)
```
?

bnavigator replied:
Install pytest-timeout

Mark-Yeatman replied:
That's the ticket!

bnavigator commented:
> Seems like the pipeline is having some issues unrelated to my changes?

Fixed. I merged current main into your PR. You have to pull your remote branch before continuing to work or rebase your commits onto the updated main branch yourself.

Mark-Yeatman commented on control/passivity.py:
```diff
@@ -1,41 +1,68 @@
+Date: July 17, 2022
 '''
 
+from msilib.schema import Error
```
My vscode extensions really like adding random imports when I mistype things.

bnavigator commented on control/tests/passivity_test.py:
```diff
@@ -54,6 +114,17 @@ def test_ispassive_edge_cases(test_input, expected):
+    D = numpy.array([[0]])
+    sys = ss(A, B, C, D)
+
+    with pytest.raises(Exception):
```
Check for the more specific exception

Mark-Yeatman replied:
Change taken.

Mark-Yeatman commented on control/passivity.py:
```diff
@@ -45,43 +71,142 @@ def ispassive(sys):
+
+    Returns
+    -------
+    bool or float: 
```
Ah, good catch, the "None" case is for when you try to get a passivity index but the problem is unfeasible.

murrayrm commented on control/passivity.py:
```diff
@@ -45,43 +71,143 @@ def ispassive(sys):
+            return False
+
+
+def getPassiveIndex(sys, index_type=None):
```
The convention in python-control is to use intercaps for class names, but not functions.  This should be `get_passive_index` (or `get_passivity_index`?).

Mark-Yeatman commented:
@bnavigator I'm still confused about why the checks are failing?

Mark-Yeatman commented:
Turns out I misread the documentation for the feedback function and got the function arguments in the wrong order. Everything should work now.

murrayrm commented:
There's some uncovered code that it would be useful to catch in the unit tests.  You can see the lines that are missing [here](https://coveralls.io/builds/51075148/source?filename=control%2Fpassivity.py).

bnavigator commented:
I have to admit that it is a bit confusing to me what nu, rho and the various instances of None mean. Especially that the meaning of the ispassive return value depends on them. So unless this is some standard convention adopted from the paper algorithms, I would rather like to see these cases implemented in separate API functions and methods. Same for `get_passivity_index()`.

Mark-Yeatman commented:
> I have to admit that it is a bit confusing to me what nu, rho and the various instances of None mean. Especially that the meaning of the ispassive return value depends on them. So unless this is some standard convention adopted from the paper algorithms, I would rather like to see these cases implemented in separate API functions and methods. Same for `get_passivity_index()`.

I'm trying to copy matlab's API here https://www.mathworks.com/help/control/ref/lti.getpassiveindex.html and https://www.mathworks.com/help/control/ref/lti.ispassive.html

bnavigator commented:
I see. The MATLAB function `isPassive` always returns the matlab equivalent of `bool` for the first return value `pf` and `R` as the passivity index if applicable. We usually mimic something like this with a `return_index` kwarg in the function.

Something like:

```
return (pf, R) if return_index else pf
```

Exact mimicry of the MATLAB function, if desired at all, should only be necessary in `control.matlab.isPassive()`

Note that the MATLAB functions claim to return a meaningful return value with both `rho` and `nu` not given which is equivalen to both `None` in your implementation.

Mark-Yeatman commented:
> if desired at all, should only be necessary in `control.matlab.isPassive'

Gotcha, I will change it around so that there's different function calls for each index type.

Mark-Yeatman commented:
@bnavigator I've split get_passivity_index() into get_output_fb_index() and get_input_ff_index().  I've also removed all the weird conditionals in ispassive(), it just has one code path now. I kept all the nu's and rho's in _solve_passivity_LMI() because it's easier to match the syntax from the source paper that way IMO.

murrayrm commented:
It looks like the passivity module is not (yet) loaded in __init__.m.  You should probably do that, so people don't have to import that passivity module explicitly (which allows `ct.ispassive(sys)`).  Also, we should update the Sphinx documentation to include `ispassive` in the list of functions that are implemented (in `doc/control.rst`).

Other than that, looks ready to go?

bnavigator commented on control/passivity.py:
```diff
@@ -45,43 +84,127 @@ def ispassive(sys):
+
+    Returns
+    -------
+    bool, float, or None: 
```
Is only bool now

Mark-Yeatman replied:
yep, good catch

bnavigator commented on control/passivity.py:
```diff
@@ -45,43 +84,127 @@ def ispassive(sys):
+    '''
+    Indicates if a linear time invariant (LTI) system is passive
+
+    Parameters
```
Lacks the description for `rho` and `nu`

Mark-Yeatman replied:
meant to remove them

bnavigator commented on control/passivity.py:
```diff
@@ -45,43 +84,127 @@ def ispassive(sys):
+        return sol[-1]
+
+
+def get_input_ff_index(sys, index_type=None):
```
`index_type` unused

Mark-Yeatman replied:
almost meant to remove it

murrayrm commented on control/passivity.py:
```diff
@@ -45,43 +84,126 @@ def ispassive(sys):
+
+
+def get_input_ff_index(sys):
+    '''
+    Returns the input feedforward passivity (IFP) index for the input system. This is the largest gain that can be 
+    placed in negative parallel interconnection with a system such that the new interconnected system is passive.
+    Parameters
```
As above.

murrayrm commented on control/passivity.py:
```diff
@@ -45,43 +84,126 @@ def ispassive(sys):
 
-    return (sol["x"] is not None)
+def get_output_fb_index(sys):
+    '''
+    Returns the output feedback passivity index for the input system. This is largest gain that can be placed in
+    positive feedback with a system such that the new interconnected system is passive.
+    Parameters
```
This first line of the docstring needs to be a short phrase that describes the function, so that the table entriex in the sphinx documentation show up correctly.  See https://numpydoc.readthedocs.io/en/latest/format.html for information (in particular, "short summary" versus "extended summary").  See [`iosys.py`](https://github.com/python-control/python-control/blob/main/control/iosys.py) for lots of examples.

Also, I think you need a blank line before Parameters for Sphinx to properly parse things.

murrayrm reviewed (commented):
More small things to fix (sorry I didn't notice these on my previous pass..).

bnavigator commented on control/passivity.py:
```diff
@@ -4,7 +4,7 @@
 
 import numpy as np
-from control import statesp as ss
+from control import statesp
```
The name collision was a friendly reminder, that you should only export your own functions when `from .passivity import *` is used: Define `__all__`.

bnavigator commented on control/passivity.py:
```diff
@@ -45,43 +84,126 @@ def ispassive(sys):
+        return sol[-1]
+
+
+def ispassive(sys):
```
But what about the cases, when you want to determine (boolean)
- input passivity (`nu != 0, rho = 0`), 
- output passivity (`nu = 0, rho != 0`)

now? One would have to use the `get_*_index()` functions. There is also no function to determine the relative passivity index.

Sorry if I created confusion earlier. I meant to make the function return values intuitive, not to remove functionality.

Mark-Yeatman replied:
You could determine input passivity with  `is_input_passive = get_input_ff_index(sys) > 0` (same for output passivity). I could add two more functions that do just that. 

Right now, I'm intentionally not fiddling with computing the relative passivity index, I need to go read the source that the Matlab documentation is citing "Sector Bounds in Stability Analysis and Control Design". If you think it should be included in this pull request, I'm totally willing to add it.

bnavigator replied:
> I could add two more functions that do just that.

Despite earlier discussions, I would vote for including that functionality into `ispassive(sys, nu, rho)`

bnavigator replied:
> Right now, I'm intentionally not fiddling with computing the relative passivity index, I need to go read the source that the Matlab documentation is citing "Sector Bounds in Stability Analysis and Control Design". If you think it should be included in this pull request, I'm totally willing to add it.

No that is fine. This can be defered to later PRs. Just raise a `NotImplemented` Exception instead of stating "This input is invalid". Same for I/O passivity index, I guess.

Mark-Yeatman replied:
> But what about the cases, when you want to determine (boolean)
> 
> * input passivity (`nu != 0, rho = 0`),
> * output passivity (`nu = 0, rho != 0`)
> 
> now? One would have to use the `get_*_index()` functions. 
> 
> Sorry if I created confusion earlier. I meant to make the function return values intuitive, not to remove functionality.

I believe I've captured this use case now.

Mark-Yeatman replied:
> No that is fine. This can be defered to later PRs. Just raise a NotImplemented Exception instead of stating "This input is invalid". **Same for I/O passivity index, I guess.**

I'm confused about the bolded part?

bnavigator replied:
Sorry, forget the bolded part. Sloppy editing by me. When there is no `get_passive_index()` catchall function you don't need  that.

But *(optional)* nice to have as a future reminder to implement it:

```python
def get_relative_index(sys):
    """Returns the relative passivity index for the system.

    (not implemented yet)
    """
    raise NotImplemented("Relative passivity index not implemented")


def get_combined_io_index(sys):
    """Returns the combined I/O passivity index for the system.

    (not implemented yet)
    """
    raise NotImplemented("Combined I/O passivity index not implemented")


def get_directional_index(sys):
    """Returns the directional passivity index for the system.

    (not implemented yet)
    """
    raise NotImplemented("Directional passivity index not implemented")

```

Mark-Yeatman replied:
Awesome, I'll take that change.

bnavigator commented on control/passivity.py:
```diff
@@ -45,43 +84,126 @@ def ispassive(sys):
+    Returns
+    -------
+    bool: 
+        The input system is passive.
```
"input"?

bnavigator commented on control/tests/passivity_test.py:
```diff
@@ -54,6 +114,26 @@ def test_ispassive_edge_cases(test_input, expected):
+    with pytest.raises(ValueError):
+        passivity.ispassive(sys)
+
+def test_rho_and_nu_are_none():
```
Is this really a forbidden case? I see it is in the current implementation, but is it from the definition? Also, what's the difference between 0 and `None` exactly?

Mark-Yeatman replied:
I'm trying to use `None` as a marker to treat the index as a variable to be maximized. Trying to maximize both the input and output passivity indices is **bilinear** matrix inequality, versus a **linear** matrix inequality. The former is **not** convex.

bnavigator replied:
Alright, you make a distinction between the parameters for `ispassive()` and `solve_passivity_LMI()`. Got it.

bnavigator commented on control/passivity.py:
```diff
@@ -45,43 +69,180 @@ def ispassive(sys):
+
+
+def ispassive(sys, ofp_index = 0, ifp_index = 0):
+    '''Indicates if a linear time invariant (LTI) system is passive.
+
+    Checks if system is passive with the given output feedback (OFP) and input feedforward (IFP)
+    passivity indices. Querying if the system is passive in the sense of V(x)>=0 and \\dot{V}(x) <= y.T*u, 
+    is equiavlent to the default case of ofp_index = 0, ifp_index = 0.
+
+    Note that computing the ofp_index and ifp_index for a system, then using both values simultaneously 
+    to as inputs to this function is not guaranteed to have an output of 'True' 
+    (the system might not be passive with both indices at the same time). For more details, see:
+        McCourt, Michael J., and Panos J. Antsaklis
+            "Demonstrating passivity and dissipativity using computational methods." 
+
+    Parameters
+    ----------
+    sys: An LTI system
+        System to be checked.
+    ofp_index: float
+        Output feedback passivity index.
+    ifp_index: float
+        Input feedforward passivity index.
+
+    Returns
+    -------
+    bool: 
+        The system is passive.
+    '''
```
"equiavlent"

The math doesn't render well:

![ispassive_spyder](https://user-images.githubusercontent.com/4623504/181465137-9c293e7d-c13a-4963-a166-5ce9c28df4e8.png)

This one does not throw any linter errors at me:

```suggestion
    r"""Indicate if a linear time invariant (LTI) system is passive.

    Checks if system is passive with the given output feedback (OFP) and input
    feedforward (IFP) passivity indices.

    Parameters
    ----------
    sys: An LTI system
        System to be checked.
    ofp_index: float
        Output feedback passivity index.
    ifp_index: float
        Input feedforward passivity index.

    Returns
    -------
    bool: 
        The system is passive.

    Notes
    -----
    Querying if the system is passive in the sense of
    
    .. math:: V(x) >= 0 \land \dot{V}(x) <= y^T u

    is equivalent to the default case of `ofp_index = 0`, `ifp_index = 0`.

    Note that computing the ofp_index and ifp_index for a system, then using
    both values simultaneously to as inputs to this function is not guaranteed
    to have an output of `True` (the system might not be passive with both
    indices at the same time).
    
    For more details, see [1].

    References
    ----------
    .. [1] McCourt, Michael J., and Panos J. Antsaklis
          "Demonstrating passivity and dissipativity using computational methods." 
    """   
```


![ispassive_spyder_new](https://user-images.githubusercontent.com/4623504/181469028-f64d21c5-52ed-45a4-9161-daca14a7ece8.png)


(Yours shows me D300 D301 D401 in spyder)
https://numpydoc.readthedocs.io/en/latest/format.html
http://www.pydocstyle.org/en/6.1.1/error_codes.html

Mark-Yeatman replied:
That is much better, thanks for doing that!

Mark-Yeatman commented:
> You actually need the `r"""` here or have to use double backslashes below.

Is there a name for what this is?

bnavigator commented:
> Is there a name for what this is?

["longstring stringliteral with raw stringprefix" or "raw triple-quoted strings"](https://docs.python.org/3/reference/lexical_analysis.html#string-and-bytes-literals)

Mark-Yeatman commented:
> > Is there a name for what this is?
> 
> ["longstring stringliteral with raw stringprefix" or "raw triple-quoted strings"](https://docs.python.org/3/reference/lexical_analysis.html#string-and-bytes-literals)
> 
> [PEP 257](https://peps.python.org/pep-0257/):
> 
> > For consistency, always use `"""triple double quotes"""` around docstrings. Use `r"""raw triple double quotes"""` if you use any backslashes in your docstrings. For Unicode docstrings, use `u"""Unicode triple-quoted strings"""`.

Noice

murrayrm commented:
@Mark-Yeatman @bnavigator This looks like it is ready to merge?  Anything else planned?

bnavigator commented:
Thanks @Mark-Yeatman! Looking forward to future pull request with further enhancements of this module.
